File: clojure/largest-series-product/src/largest_series_product.clj
(ns largest-series-product)

(defn- to-int [c] (- (int c) (int \0)))

(defn- valid? [digits] (every? #(<= (int \0) (int %) (int \9)) digits))

(defn largest-product [n digits]
  {:pre ((<= 0 n (count digits))
         (valid? digits))}
  (if (= n 0) 1
    (->> digits
         (map to-int)
         (partition n 1)
         (map #(apply * %))
         (apply max))))

File: javascript/largest-series-product/largest-series-product.js
function Series (num) {
    'use strict';

    this._num = num;
}

function partition (arr, len, offset) {
    'use strict';

    if (arr.length < len) {
        throw new Error('Slice size is too big.');
    }
    var ret = [];
    for (var i = 0; i + len <= arr.length; i += offset) {
        ret.push(arr.slice(i, i + len));
    }
    return ret;
}

Series.prototype.largestProduct = function (len) {
    'use strict';

    if (len < 0 || /[^0123456789]/.test(this._num)) {
        throw new Error('Invalid input.');
    }
    if (len === 0) {
        return 1;
    }
    var nums = this._num.split('');
    var parts = partition(nums, len, 1);
    return parts.reduce(function (pval, cval) {
        var product = 1;
        for (var i = 0; i < cval.length; i ++) {
            product = product * parseInt(cval[i]);
        }
        return Math.max(pval, product);
    }, 0);
};

module.exports = Series;

